File: scripts/reprocess_documents.py
#!/usr/bin/env python3
"""
Direct document reprocessing using Python (avoids Node 25 OOM bug).
Reads raw text from S3, chunks it, embeds with Titan V2, stores in pgvector.

Usage: python3 scripts/reprocess_documents.py
"""
import os
import json
import boto3
import psycopg2
from dotenv import load_dotenv

load_dotenv()

DB_URL = os.environ['DATABASE_URL']
REGION = os.environ.get('AWS_REGION', 'us-east-1')
S3_BUCKET = os.environ.get('S3_BUCKET_NAME', 'fundlens-documents-dev')

s3 = boto3.client('s3', region_name=REGION)
bedrock = boto3.client('bedrock-runtime', region_name=REGION)


def get_s3_text(key):
    resp = s3.get_object(Bucket=S3_BUCKET, Key=key)
    return resp['Body'].read().decode('utf-8')


def generate_embedding(text, dims=1024):
    """Generate embedding using Titan V2. Returns list of floats."""
    truncated = text[:32000]
    resp = bedrock.invoke_model(
        modelId='amazon.titan-embed-text-v2:0',
        contentType='application/json',
        accept='application/json',
        body=json.dumps({
            'inputText': truncated,
            'dimensions': dims,
            'normalize': True,
        }),
    )
    result = json.loads(resp['body'].read())
    return result['embedding']


def chunk_text(raw_text, max_chars=2400, overlap=100):
    """Split text into chunks with overlap."""
    chunks = []
    idx = 0
    start = 0
    while start < len(raw_text):
        end = min(start + max_chars, len(raw_text))
        if end < len(raw_text):
            seg = raw_text[start:end]
            bp = max(seg.rfind('. '), seg.rfind('\n'))
            if bp > max_chars * 0.5:
                end = start + bp + 1
        content = raw_text[start:end].strip()
        if len(content) > 50:
            chunks.append({'idx': idx, 'content': content, 'section': 'narrative'})
            idx += 1
        start = end - overlap
        if start >= len(raw_text):
            break
    return chunks


def extract_text_from_pdf(s3_key):
    """Download PDF from S3 and extract text using PyPDF2."""
    import io
    try:
        from PyPDF2 import PdfReader
    except ImportError:
        print("  Installing PyPDF2...")
        import subprocess
        subprocess.check_call(['pip3', 'install', 'PyPDF2', '-q'])
        from PyPDF2 import PdfReader

    resp = s3.get_object(Bucket=S3_BUCKET, Key=s3_key)
    pdf_bytes = resp['Body'].read()
    reader = PdfReader(io.BytesIO(pdf_bytes))
    text_parts = []
    for page in reader.pages:
        text_parts.append(page.extract_text() or '')
    return '\n\n'.join(text_parts)


def process_document(conn, doc):
    doc_id = doc['document_id']
    print(f"\n=== {doc['file_name']} ({doc_id}) ===")
    print(f"  status={doc['status']}, mode={doc['processing_mode']}, chunks={doc['chunk_count']}")

    # Step 1: Get raw text
    raw_text = ''
    if doc['raw_text_s3_key']:
        print(f"  Reading raw text from S3...")
        raw_text = get_s3_text(doc['raw_text_s3_key'])
    elif doc['s3_key']:
        print(f"  Extracting text from PDF...")
        raw_text = extract_text_from_pdf(doc['s3_key'])
        # Store raw text to S3
        raw_key = f"extracted/{doc['tenant_id']}/{doc['deal_id']}/{doc_id}/raw_text.txt"
        s3.put_object(Bucket=S3_BUCKET, Key=raw_key, Body=raw_text.encode('utf-8'), ContentType='text/plain')
        cur = conn.cursor()
        cur.execute(
            "UPDATE intel_documents SET raw_text_s3_key = %s, status = 'queryable', "
            "processing_mode = 'long-context-fallback', page_count = %s, updated_at = NOW() "
            "WHERE document_id = %s::uuid",
            (raw_key, max(1, len(raw_text) // 3000), doc_id)
        )
        conn.commit()
        print(f"  Extracted {len(raw_text)} chars")
    else:
        print("  ❌ No text source")
        return False

    print(f"  Raw text: {len(raw_text)} chars")

    # Step 2: Clean existing chunks
    cur = conn.cursor()
    cur.execute("DELETE FROM intel_document_chunks WHERE document_id = %s::uuid", (doc_id,))
    conn.commit()

    # Step 3: Chunk
    chunks = chunk_text(raw_text)
    print(f"  Created {len(chunks)} chunks")
    if not chunks:
        return False

    # Step 4: Embed and index one at a time
    indexed = 0
    for i, chunk in enumerate(chunks):
        try:
            embedding = generate_embedding(chunk['content'])
            emb_str = '[' + ','.join(str(x) for x in embedding) + ']'

            cur = conn.cursor()
            cur.execute(
                "INSERT INTO intel_document_chunks "
                "(id, document_id, tenant_id, deal_id, chunk_index, content, section_type, "
                "token_estimate, embedding, created_at) "
                "VALUES (gen_random_uuid(), %s::uuid, %s::uuid, %s::uuid, %s, %s, %s, %s, %s::vector, NOW())",
                (doc_id, doc['tenant_id'], doc['deal_id'],
                 chunk['idx'], chunk['content'], chunk['section'],
                 len(chunk['content']) // 4, emb_str)
            )
            conn.commit()
            indexed += 1

            if (i + 1) % 5 == 0 or i == len(chunks) - 1:
                print(f"  Indexed {indexed}/{len(chunks)} chunks")
        except Exception as e:
            print(f"  ⚠️ Chunk {i} failed: {e}")
            conn.rollback()

    # Step 5: Update document
    cur = conn.cursor()
    cur.execute(
        "UPDATE intel_documents SET processing_mode = 'fully-indexed', chunk_count = %s, "
        "status = 'queryable', error = NULL, updated_at = NOW() WHERE document_id = %s::uuid",
        (indexed, doc_id)
    )
    conn.commit()
    print(f"  ✅ DONE: {indexed} chunks indexed")
    return True


def main():
    # Strip Prisma-specific params from DATABASE_URL
    db_url = DB_URL.split('?')[0]
    conn = psycopg2.connect(db_url, sslmode='require')
    print("Connected to DB")

    cur = conn.cursor()
    cur.execute(
        "SELECT document_id, file_name, status, processing_mode, chunk_count, "
        "tenant_id, deal_id, raw_text_s3_key, s3_key, file_type "
        "FROM intel_documents ORDER BY created_at"
    )
    columns = [desc[0] for desc in cur.description]
    docs = [dict(zip(columns, row)) for row in cur.fetchall()]
    print(f"Found {len(docs)} documents")

    for doc in docs:
        try:
            process_document(conn, doc)
        except Exception as e:
            print(f"  ❌ Failed: {e}")
            conn.rollback()

    # Final status
    print("\n========== FINAL STATUS ==========")
    cur = conn.cursor()
    cur.execute(
        "SELECT file_name, status, processing_mode, chunk_count, metric_count "
        "FROM intel_documents ORDER BY created_at"
    )
    for row in cur.fetchall():
        icon = '✅' if row[2] == 'fully-indexed' and (row[3] or 0) > 0 else '⚠️'
        print(f"{icon} {row[0]}: status={row[1]}, mode={row[2]}, chunks={row[3]}")

    conn.close()


if __name__ == '__main__':
    main()

File: scripts/reprocess-documents.js
/**
 * Reprocess documents: clean up partial enrichment data, then re-trigger
 * the full pipeline (Phase A + Phase B) one document at a time.
 *
 * Usage: node scripts/reprocess-documents.js
 */
require('dotenv').config();
const { Client } = require('pg');

const DB_URL = process.env.DATABASE_URL;
const BASE_URL = 'http://localhost:3000';
const TENANT_ID = '00000000-0000-0000-0000-000000000000';

// Craft a minimal JWT that passes decode-only mode
function makeDevToken(tenantId) {
  const header = Buffer.from(JSON.stringify({ alg: 'none', typ: 'JWT' })).toString('base64url');
  const payload = Buffer.from(JSON.stringify({
    sub: 'reprocess-script',
    'custom:tenant_id': tenantId,
    'custom:tenant_role': 'admin',
    'custom:tenant_slug': 'default',
    username: 'admin@fundlens.dev',
  })).toString('base64url');
  return `${header}.${payload}.nosig`;
}

async function main() {
  const client = new Client({ connectionString: DB_URL, ssl: { rejectUnauthorized: false } });
  await client.connect();
  console.log('Connected to DB');

  const token = makeDevToken(TENANT_ID);
  const authHeaders = {
    'Content-Type': 'application/json',
    'Authorization': `Bearer ${token}`,
  };

  // Get all documents
  const { rows: docs } = await client.query(
    `SELECT document_id, file_name, status, processing_mode, chunk_count, 
            tenant_id, deal_id, s3_key, file_type
     FROM intel_documents ORDER BY created_at`
  );

  console.log(`Found ${docs.length} documents to reprocess\n`);

  for (let i = 0; i < docs.length; i++) {
    const doc = docs[i];
    console.log(`\n[${i + 1}/${docs.length}] ${doc.file_name} (${doc.document_id})`);
    console.log(`  Current: status=${doc.status}, mode=${doc.processing_mode}, chunks=${doc.chunk_count}`);

    // Step 1: Clean up partial vision extractions (keep headline extractions from Phase A)
    const delVision = await client.query(
      `DELETE FROM intel_document_extractions WHERE document_id = $1::uuid AND source_layer = 'vision'`,
      [doc.document_id]
    );
    console.log(`  Cleaned ${delVision.rowCount} partial vision extractions`);

    // Step 2: Clean up any partial chunks
    try {
      const chunkDel = await client.query(
        `DELETE FROM intel_document_chunks WHERE document_id = $1::uuid`,
        [doc.document_id]
      );
      console.log(`  Cleaned ${chunkDel.rowCount} partial chunks`);
    } catch (e) {
      console.log(`  No chunks to clean`);
    }

    // Step 3: Delete headline extractions too (they'll be re-created)
    await client.query(
      `DELETE FROM intel_document_extractions WHERE document_id = $1::uuid`,
      [doc.document_id]
    );

    // Step 4: Reset document to uploading so upload-complete will re-run full pipeline
    await client.query(
      `UPDATE intel_documents SET
        status = 'uploading',
        processing_mode = NULL,
        chunk_count = 0,
        metric_count = 0,
        raw_text_s3_key = NULL,
        document_type = NULL,
        company_ticker = NULL,
        company_name = NULL,
        page_count = NULL,
        error = NULL,
        retry_count = 0,
        kb_sync_status = 'pending',
        updated_at = NOW()
      WHERE document_id = $1::uuid`,
      [doc.document_id]
    );
    console.log(`  Reset to clean uploading state`);

    // Step 5: Trigger via upload-complete endpoint
    console.log(`  Triggering full pipeline (Phase A + B)...`);
    try {
      const resp = await fetch(`${BASE_URL}/api/documents/${doc.document_id}/upload-complete`, {
        method: 'POST',
        headers: authHeaders,
      });
      if (resp.ok) {
        const result = await resp.json();
        console.log(`  Phase A complete: type=${result.documentType}, company=${result.companyName}, ticker=${result.ticker}`);
        console.log(`  Phase B (enrichment) running in background...`);
      } else {
        const text = await resp.text();
        console.log(`  ❌ API returned ${resp.status}: ${text}`);
        continue;
      }
    } catch (e) {
      console.log(`  ❌ API call failed: ${e.message}`);
      continue;
    }

    // Step 6: Poll for enrichment completion (every 5s, max 4 min)
    let attempts = 0;
    const maxAttempts = 48; // 4 minutes
    while (attempts < maxAttempts) {
      await new Promise(r => setTimeout(r, 5000));
      const { rows } = await client.query(
        `SELECT status, processing_mode, chunk_count, error FROM intel_documents WHERE document_id = $1::uuid`,
        [doc.document_id]
      );
      const current = rows[0];
      attempts++;

      if (current.processing_mode === 'fully-indexed' && current.chunk_count > 0) {
        console.log(`  ✅ DONE: ${current.chunk_count} chunks indexed, mode=fully-indexed`);
        break;
      }
      if (current.error && current.error.includes('permanently')) {
        console.log(`  ❌ FAILED: ${current.error}`);
        break;
      }
      if (attempts % 4 === 0) {
        console.log(`  ... ${attempts * 5}s elapsed — mode=${current.processing_mode}, chunks=${current.chunk_count}, error=${current.error || 'none'}`);
      }
    }

    if (attempts >= maxAttempts) {
      // Check final state
      const { rows } = await client.query(
        `SELECT status, processing_mode, chunk_count, error FROM intel_documents WHERE document_id = $1::uuid`,
        [doc.document_id]
      );
      const final = rows[0];
      console.log(`  ⚠️ Timed out — final: mode=${final.processing_mode}, chunks=${final.chunk_count}, error=${final.error || 'none'}`);
    }

    // Wait between documents to let GC recover
    if (i < docs.length - 1) {
      console.log(`  Cooling down 10s before next document...`);
      await new Promise(r => setTimeout(r, 10000));
    }
  }

  // Final status
  console.log('\n\n========== FINAL STATUS ==========');
  const { rows: final } = await client.query(
    `SELECT document_id, file_name, status, processing_mode, chunk_count, metric_count, kb_sync_status, error
     FROM intel_documents ORDER BY created_at`
  );
  for (const d of final) {
    const icon = d.processing_mode === 'fully-indexed' && d.chunk_count > 0 ? '✅' : '⚠️';
    console.log(`${icon} ${d.file_name}`);
    console.log(`   status=${d.status}, mode=${d.processing_mode}, chunks=${d.chunk_count}, metrics=${d.metric_count}, kb=${d.kb_sync_status}`);
    if (d.error) console.log(`   error: ${d.error}`);
  }

  await client.end();
}

main().catch(e => { console.error(e); process.exit(1); });
